Paper: Experience report of physics-informed neural networks in fluid simulations: pitfalls and frustration

piyueh opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@stargaser<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@skillmananalyticsllc<!--end-reviewers-list-->

piyueh committed:
add the first draft

labarba committed:
edits abstract and intro

labarba committed:
edits to methods section

labarba committed:
edits Taylor-Green section

labarba committed:
edits to cylinder section

labarba committed:
edits conclusion

piyueh committed:
typo: larage -> large

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

piyueh committed:
fix citation key

piyueh committed:
change the title

piyueh commented:
@scoobies help

scoobies commented:
Hello @piyueh, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Add to this issue's reviewers list
@scoobies add @username to reviewers

# Remove from this issue's reviewers list
@scoobies remove @username from reviewers

# Assign a user as the editor of this submission
@scoobies assign @username as editor

# Remove the editor assigned to this submission
@scoobies remove editor

# Add a user to this issue's assignees list
@scoobies add assignee: @username

# Remove a user from this issue's assignees list
@scoobies remove assignee: @username

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: needs-review. Remove labels: unready, needs-more-review, pending-comment, ready
@scoobies mark needs review

# Label issue with: needs-more-review. Remove labels: unready, needs-review, pending-comment, ready
@scoobies mark needs more review

# Label issue with: pending-comment. Remove labels: unready, needs-review, needs-more-review, ready
@scoobies mark pending comment

# Label issue with: ready. Remove labels: unready, needs-review, needs-more-review, pending-comment
@scoobies mark ready

# Label issue with: unready. Remove labels: needs-review, needs-more-review, pending-comment, ready
@scoobies mark not ready

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

piyueh commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper barbagroup-barbagroup-scipy-2022 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

piyueh committed:
fix typos in tgv section

piyueh committed:
fix typos in cylinder section

piyueh committed:
fix typos in the last section

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"barbagroup-barbagroup-scipy-2022"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmplbunc416'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: PiYueh_Chuang\n*** Warning: /tmp/tmplbunc416/publisher/../output/PiYueh_Chuang/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/barbagroup-barbagroup-scipy-2022.pdf", "build_status"=>"success", "build_timestamp"=>"27/05 21:47"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @stargaser as editor

scoobies commented:
Assigned! @stargaser is now the editor

piyueh committed:
update with reporducibility package

piyueh committed:
correct math and update figures

piyueh committed:
fix more typos and redundant spaces

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

deniederhut commented:
@scoobies add @skillmananalyticsllc to reviewers

scoobies commented:
@skillmananalyticsllc added to the reviewers list!

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+      & r_{5}(\vec{x}, t; \Theta) \equiv  G^{p}-p_\Gamma = 0,\ \forall \vec{x} \in \Gamma \\
+   \end{aligned}
+
+And the set of desired parameter, :math:`\Theta=\theta`, is the common zero root of the all residuals.
```
I assume you mean "all the residuals"?

labarba replied:
Done.

labarba replied:
[ea26241](https://github.com/barbagroup/scipy_proceedings/commit/ea262418b8efb7cf2eaab4398587e4d67afaa50a)

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+2.1. The PINN method
+++++++++++++++++++++
+
+The basic form of the PINN method starts from approximating :math:`\vec{U}` and :math:`p` with a neural network:
```
This is just a suggestion, but readers (including me) would find it useful to see references to earlier work to define/explore possibilities for the PINN method. It seems like you present it as equally established/well defined to the governing equations, but this is recent enough it seems there should be cites.

labarba replied:
Done. [62e58e1](https://github.com/barbagroup/scipy_proceedings/commit/62e58e16fce6b85fccbabe6cfc4a3ce5df3ed197)

labarba replied:
We can add one more reference to a review-like article. @piyueh

piyueh replied:
Done in https://github.com/barbagroup/scipy_proceedings/commit/afa3d036bf6098e21ecf1294b9a4e6e2b7ce3690

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+And the set of desired parameter, :math:`\Theta=\theta`, is the common zero root of the all residuals.
+
+The derivatives of :math:`G` with respect to :math:`\vec{x}` and :math:`t` are usually obtained using automatic differentiation. 
+Nevertheless, it is possible to use analytical derivatives when the chosen network architecture is simple enough, as reported by early-day literature (:cite:`lagaris_artificial_1998,Li2003`).
```
Has anyone done this in more recent work? If so, again would be nice to cite this for readers who wish to learn more.

labarba replied:
We're not aware. The analytical derivatives can be obtained for simple network architectures, which were used in early works. In recent works, architectures are more complex and analytical derivatives are intractable.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+The derivatives of :math:`G` with respect to :math:`\vec{x}` and :math:`t` are usually obtained using automatic differentiation. 
+Nevertheless, it is possible to use analytical derivatives when the chosen network architecture is simple enough, as reported by early-day literature (:cite:`lagaris_artificial_1998,Li2003`).
+
+If residuals in (:ref:`eq:residuals`) are not complicated, and if the number of the parameters, :math:`N_\Theta`, is small enough, we may numerically find the zero root by solving a system of :math:`N_\Theta` nonlinear equations generated from a proper set of :math:`N_\Theta` spatial-temporal points.
```
Is "proper set" a technical term? If not, I might recommend rewriting this as "suitable set"

labarba replied:
Done. [0a74624](https://github.com/barbagroup/scipy_proceedings/commit/0a746241b7957b97b33ede69d76a9465f7c24961)

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Moreover, we do not even know if such a zero root exists for equations in (:ref:`eq:residuals`).
+
+Instead, in PINN, the condition is relaxed.
+We do not seek the zero root of (:ref:`eq:residuals`) but just hope to find a set of parameters that make the residuals as close to zero.
```
"as close to zero" -> "sufficiently close to zero"

labarba replied:
Fixed. [66dc35b](https://github.com/barbagroup/scipy_proceedings/commit/66dc35b00c425137fa28d349cabd86b557d92254)

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+
+   \theta = \operatorname*{arg\,min}\limits_{\Theta} r(\vec{x}, t; \Theta)\,\ \forall \left\{\begin{array}{l}x \in \Omega \\ t\in[0, T]\end{array}\right.
+
+This poses a fundamental difference between the PINN method and the traditional CFD schemes, causing the PINN method potentially more difficult to achieve the same accuracy as the traditional schemes.
```
This sentence needs a rewrite. I am not entirely clear on what your meaning is.

labarba replied:
This is actually commenting on a commit that was not the latest. The text now reads: "making it potentially more difficult for the PINN method to achieve the same accuracy as the traditional schemes."

piyueh replied:
To clarify, the sentence addressed by the reviewer was from commit https://github.com/barbagroup/scipy_proceedings/commit/aac7497b229a9cacf0895408b7403cc9cb00505b and had been revised in https://github.com/barbagroup/scipy_proceedings/commit/60db1fc3400f4bee497d484dc6dd78d901d52e3f.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Alternatively, Quasi-Newton methods may work but only when :math:`N_\Theta` is small enough.
+
+However, even though the equation (:ref:`eq:objective`) may be solvable, it is still a significantly expensive task.
+Compared to regular data-driven learning applications, due to the use of automatic differentiation to evaluate the derivatives of :math:`G` with respect to :math:`\vec{x}` and :math:`t`, the computational graph becomes much larger.
```
If you have some back of the envelope estimations of "much larger" that would be interesting to know.

labarba replied:
The passage explains why the computational graph is much larger (multiple back-propagation passes), but one had to keep reading and interpret the text. Suppose a simple back-propagation pass takes N work, if we have to take another derivative that might double the work, and so on. But we can't give a quantitative estimate of how much the computational graph is going to grow with each derivative. @piyueh can likely say it better.

We have now re-written the passage to make it more clear. See [2db67dd](https://github.com/barbagroup/scipy_proceedings/commit/2db67dd1ebb2268c4a13075f131ab5d1bdbc198a)

piyueh replied:
Here's a made-up example: if the size of a network's computational graph is N, and if AD (automatic differentiation) with respect to input data through back-propagation will double the graph size, we have the following math:
* the 1st-order derivative through AD: the graph size is 2N
* the 2nd-order derivative through AD on top of the 1st-order derivative's graph: the graph size is 4N
* the loss's (which contains the graph of the 2nd-order derivative) gradients with respect to parameters: the graph size is 8N.

So the final computational graph is 8 times larger than the original network. 

In reality, the loss term in the PINN method contains more than just the graph of the 2nd-order derivative. For example, in the Navier-Stokes equations, the loss term includes the linear and nonlinear combinations of the graphs from the 0th, 1st, and 2nd-order derivatives. So even though "AD doubles the graph size" is totally a made-up assumption, I think it's reasonable to use "much larger" here.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+
+With this workflow in mind, it should be easy to see the analogy between PINN and conventional numerical methods.
+Aside from using much more complicated approximate solutions, the major difference lies in how to determine the unknown parameters in the approximate solutions.
+While traditional methods solve the zero-residual conditions, PINN relies on searching the minimal residuals.
```
Isn't there also quite a fundamental difference in paradigms? You indicate here that PINNs are trained on data from traditional CFD models. This seems worth recapitulating here as well even if it does not go to the training, it goes to the external validity and potentially unaccounted for differences in the training data/overall training paradigm.

labarba replied:
The PINN method is _not_ trained on data from traditional CFD models. Can you explain why you get this perception so we can look for a way to be more clear in the paper? Of course our methods section is very short (this is a conference paper and we have a limitation of 8 pages), but we would like to make every effort to clear this confusion.

piyueh replied:
Let me see if this explanation is clearer. Another way to describe PINN using the typical data-driven machine learning framework is as follows:

* the inputs of the network are just temporal and spatial coordinates, e.g., (x, y) and t for a 2D heat conduction problem
* the outputs of the network are the physical quantities of our interest, e.g., the predicted temperature at a given (x, y) and t for a heat conduction problem
* the objective/loss functions are the partial differential equations that constrain the physical quantities, e.g., the [2D heat equation](https://en.wikipedia.org/wiki/Heat_equation), which constrains (i.e., regulates, or governs) the temperature at a given (x, y) and t.

Because the loss functions are governing equations (rather than the differences between predictions and the true answers), they don't need any "data". So the PINN method does not need and is not trained on data from traditional CFD models. We may be using the term "training data" throughout the paper. Nevertheless, they actually mean "spatio-temporal points" for the networks' inputs.

piyueh replied:
I added a paragraph try to explain this. See https://github.com/barbagroup/scipy_proceedings/commit/b4bcf999fc4425407688ac6e62d9ed47c90090e4

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+
+The neural network used in the PINN solver is a fully-connected neural network with 6 hidden layers and 256 neurons per layer.
+The activation functions are SiLU (:cite:`hendrycks_gaussian_2016`).
+We used Adam for optimization, and its initial parameters are the defaults from PyTorch.
```
Didn't read below yet, but did you explore cyclical learning rates?

labarba replied:
Not in the results presented here. @piyueh has been recently looking into this, but it is out of scope for this paper.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+All cases were trained up to 1 million iterations.
+Note that the parallelization was done with weak scaling, meaning increasing the number of GPUs would not reduce the workload of each GPU.
+Instead, increasing the number of GPUs would increase the total and per-iteration numbers of training points. 
+Therefore, our expected outcome was that all cases required about the same wall time to finish, while the residual from using 8 GPUs would converge the fast.
```
"the fast" -> "the fastest"

labarba replied:
This had already been fixed in a later commit than the comment here was attached to.

piyueh replied:
The issue addressed by the reviewer was from the very first commit of this PR. And the issue had been fixed in a later commit and does not exist now.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+++++++++++++
+
+Figure :ref:`fig:tgv-pinn-training-convergence` shows the convergence history of the total residuals (equation (:ref:`eq:total-residual`)).
+Using more GPUs did not accelerate the convergence, unlike what we expected previously.
```
Can you explain why you expected more GPUs to accelerate the convergence (perhaps with details of the weak scaling you reference in the previous section)?

labarba replied:
In weak scaling, more GPUs means more spatio-temporal points, which we expected to produce faster learning, because the gradient estimate used in gradient descent would be more accurate. We will add some more explanation in the method section. Stay tuned.
@piyueh : post here when done.

piyueh replied:
Weak scaling means increasing the number of GPUs without changing the workload of each GPU. So if we have N training points per batch when using 1 GPU, then we will have 8N training points per batch when using 8 GPUs. And as described in the above comment, we assumed using more training points per batch would help the training with gradient-descent methods.

Will add some more explanation in the paper.

labarba replied:
Added a few words to the sentence, to make more clear. See [2265597](https://github.com/barbagroup/scipy_proceedings/commit/226559767eb94d3d0afe77e244fa537f0120e0f5)

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+
+Figure :ref:`fig:tgv-pinn-training-convergence` shows the convergence history of the total residuals (equation (:ref:`eq:total-residual`)).
+Using more GPUs did not accelerate the convergence, unlike what we expected previously.
+All cases converged as a similar rate.
```
at a similar rate

labarba replied:
Also fixed (note the comments are attached to a commit that was not the latest on the date the review happened; unclear why).

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Figure :ref:`fig:tgv-pinn-training-convergence` shows the convergence history of the total residuals (equation (:ref:`eq:total-residual`)).
+Using more GPUs did not accelerate the convergence, unlike what we expected previously.
+All cases converged as a similar rate.
+Though without a quantitative criterion or justification, we considered that further training would not improve the accuracy.
```
Perhaps you will do so further down, but I find it surprising that you did not at this point run some kind of grid search on the training parameters you specify at the start for your NN. This can sometimes make a tremendous difference.

labarba replied:
As you saw in the conclusions, the point is that with the PINN method being so costly, doing a parameter sweep would make it even more so!

piyueh replied:
Maybe we can add a statement describing where we got these hyperparameters?

Hyperparameter optimization is too expensive for this type of application. Many machine learning applications can afford a significant computational cost because once the model is well trained, the model can be used in many similar applications or for a long time. However, for many CFD applications in computer-aided engineering settings, the trained model is only used to predict the flow at one specified flow condition. So hyperparameter optimization may not be worth it due to its price-performance ratio. As indicated in the later discussion, cost and performance are important for engineering.

However, the major point of this paper is to show what we have done and how the results look like. So if more people are willing to share their results with different hyperparameters, regardless of how good or bad the results are, I think all our attempts can be deemed as a collective effort of hyperparameter optimization. That's why when mainstream papers only share what was working well, we would want to share what was not.

piyueh replied:
I added some words to show this paper is to report our computational experiments (https://github.com/barbagroup/scipy_proceedings/commit/f676e743746d29dfabd624f2409adef5140826ff): we share our experiment configurations, the results, and our subjective thinking. We did not aim to resolve or tackle any problem.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Figure :ref:`fig:tgv-sim-time-errors` shows the prediction errors with respect to :math:`t`.
+We only present the error of :math:`u` velocity as those of :math:`v` and :math:`p` are similar.
+The accuracy of the PINN solver is similar to that of the :math:`16 \times 16` simulation with PetIBM.
+Using more GPUs, which implies more training points, does not improve the accuracy.
```
Can you say anything qualitatively/quantitatively about the relative distribution of errors?

labarba replied:
The spatial distribution of errors is not informative because the output of the PINN method also includes the time axis: at what time snapshot would we look at the errors? This also applies to the suggestion below about a visualization of errors.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+The accuracy of the PINN solver is similar to that of the :math:`16 \times 16` simulation with PetIBM.
+Using more GPUs, which implies more training points, does not improve the accuracy.
+
+Regardless of the magnitudes, the trends of errors with respect to :math:`t` are similar for both PINN and PetIBM.
```
Ah you get to this. Perhaps a visualization?

piyueh replied:
Figure 4 should be representative of temporal errors. Regarding spatial errors, we currently don't see what information we can draw from a spatial error contour. Also, we are not sure at what time snapshot should we present the spatial error. Given the page limit of the SciPy proceeding, we refrained from adding figures that we didn't have much to discuss. If there's anything worth discussing from the spatial error contours, please let us know.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Figure :ref:`fig:tgv-run-time-errors` shows the computational cost of PINN and PetIBM in terms of the desired accuracy versus the required wall time.
+We only showed the PINN results of 8 A100 GPUs on this figure.
+We believe this type of plot may help evaluate the computational cost in engineering.
+According to the figure, for example, achieving an accuracy of :math:`10^{-2}` at :math:`t=2` requires less than 20 seconds for PetIBM with 1 K40 and 6 CPU cores, but it requires more than 8 hours with at least 1 A100 GPU.
```
I would register some skepticism here since you did not justify or calibrate your hyper parameters. I would suggest you either include some experiments on hyper parameter optimization or explicitly limit your language here to the case of not doing such an investigation.

labarba replied:
One of the points we want to make in this paper is that to improve the results, one would need to search for better parameters by systematic re-execution of the model. Already one calculation with PINN is orders of magnitude more costly than a traditional CFD simulation. It would be even more costly to run over and over again to find the "best" result.

piyueh replied:
This paragraph and statement merely describe what we saw from our results. We didn't claim something like "under the PINN's best performance" in our statements. We refrained from claiming and discussing anything in the paragraphs of the result section: we just described the results. The results are naturally bounded to the specific configuration stated in section 3.2. So I'm not sure what the issue is in this statement. We provided a repro-pack for interested readers to reproduce all results and figures in the result sections: https://github.com/barbagroup/scipy-2022-repro-pack. (You were reviewing the earliest commit, which does not reflect our final submission, so some information, such as the repro-pack, might be missing from the version you reviewed.)

That being said, even if we do a hyperparameter tuning and get, for example, an optimal learning rate, figure 5 and the associated description are not affected because they are not bounded to the said new learning rate.

On the other hand, I agree we could include statements about how/why we chose these hyperparameters and a disclaimer that these hyperparameters are not optimal.

piyueh replied:
See the added paragraph in section 3.4 in https://github.com/barbagroup/scipy_proceedings/commit/f676e743746d29dfabd624f2409adef5140826ff

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Taylor-Green vortex serves as a good benchmark case because it reduces the number of required residual constraints: residuals :math:`r_4` and :math:`r_5` are excluded from :math:`r` in equation :ref:`eq:total-residual`.
+This makes the optimizer can focus only on the residuals of initial conditions and the Navier-Stokes equations.
+
+Using more GPUs (thus using more training points) did not speed up the convergence may indicate that the per-iteration number of points on a single GPU is already big enough.
```
I am a little confused here by what you mean by "more training data". I had the impression you used more spatial-temporal points with more GPUs but is this within the same output simulation? Did you simulate multiple scenarios as distinct training points? I suppose I realize at this point I'm a bit fuzzy as to what the axes of your input data were - is this a single time series for a single geospatial setup? So with more GPUs you simply have a denser grid? perhaps add some language here or earlier to clarify.

labarba replied:
In the PINN method, "training data" are spatio-temporal points. We do use more points when adding GPUs. In traditional CFD methods, when we use more points (more "refined mesh") we get higher accuracy. In PINN method, we expect faster learning, i.e., converge to the minimum faster.

There is not time series here, but with more GPUs we do have more spatio-temporal points (denser grid in space-time).

@piyueh will look into clarifying the method section; we'll add to this.

piyueh replied:
Done. See https://github.com/barbagroup/scipy_proceedings/commit/b4bcf999fc4425407688ac6e62d9ed47c90090e4.

I would like to add some remarks about spatial-temporal points here, though they may be irrelevant. In PINN, time and space do not seem to be treated differently. So it's not like classic time series problems. (For example, in parabolic and hyperbolic problems, the data at a given time can only be affected by previous time, while the data at a spatial location may be affected by the neighbors from all directions.) So beyond the constraints from the governing equations, I _think_ temporal and spatial coordinates are treated indistinguishably in PINN.

SkillmanAnalyticsLLC replied:
Thank you. This helps a lot.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+The low accuracy to some degree was not surprising.
+Recall the theory in section 2.
+The PINN method only seeks the minimal residual on the total residual's hyperplane.
+It does not try to find the zero root of the hyperplane and does not even care whether such a zero root exists.
```
Are there people writing about architectures where a zero root is built into the model somehow?

labarba replied:
We don't know. But this is not how PINN works.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+It makes sense the residual may be hard to close to zero, meaning it is hard to make error small.
+
+Regarding the performance result in figure :ref:`fig:tgv-run-time-errors`, we would like to avoid interpreting the result as one solver being better than the other one.
+The proper conclusion drawn from the figure should be as follows: when using the PINN solver as a CFD simulator for a specific flow condition, PetIBM outperforms the PINN solver.
```
As I mentioned earlier, I would also add to this "in the case where extensive hyper parameter optimization is not conducted"

piyueh replied:
See the added paragraph in section 3.4 in https://github.com/barbagroup/scipy_proceedings/commit/f676e743746d29dfabd624f2409adef5140826ff

SkillmanAnalyticsLLC replied:
thank you.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+As stated in section 1, the PINN method can solve flows under different flow parameters in one run—a capability that PetIBM does not have.
+The performance result in figure :ref:`fig:tgv-run-time-errors` only considers a limited aspect of the PINN solver.
+
+One issue in this case study was fairly comparing the PINN solver and PetIBM, especially when investigating the accuracy versus the workload/problem size or time-to-solution versus problem size.
```
This is likely not actionable by any of us, but I wonder whether there have been CFD competitions, say on Kaggle, or whether you might take this opportunity to propose one. Arguably the best way to match up methods would be to have a competition where you have lots of people incentivized to look creatively for solutions. Then you could look at say top N solutions and compare computational resources and relative error rate. That would be another way to take a look at this problem (albeit not to draw final conclusions)

piyueh replied:
I agree. And I believe this is workable (though will not be in this paper anyway.). Probably this has to be done in the CFD community. CFD community in the past sometimes organized workshops for a similar purpose.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+----------------------------------------------------
+
+This case study shows what really made us frustrated: a 2D cylinder flow at Reynolds number :math:`Re=200`.
+We failed to even produce a solution that qualitatively captures the key physical phenomenon of this flow: vortex shedding.
```
So here is where I (and likely other readers) could use some background. I have the impression that traditional CFD has been driven by the idea of qualitatively capturing correct observed behaviors. That is a fundamentally different development paradigm from PINN which is trying to reproduce the data from CFD. Arguably it is unsurprising that PINN can easily fail to reproduce qualitatively important behaviors because it is not training to do so. Rather it is training to minimize overall error. However, I wonder whether you explored work looking at how you can rewrite your loss function to encourage it to produce the expected qualitative behaviors?

piyueh replied:
I think here is a misunderstanding. I hope the revision (https://github.com/barbagroup/scipy_proceedings/commit/b4bcf999fc4425407688ac6e62d9ed47c90090e4) can clear the misunderstanding.

The PINN method's purpose in this paper is not to reproduce the data from CFD. It is meant to function as a drop-in replacement of a traditional CFD solver. They both solve the Navier-Stokes equations without users feeding any ground truth (i.e., no experimental/observation data or simulation data from other tools). Both the PINN solver and the traditional CDF solver were given the same flow conditions (e.g., the flow velocity at t=0 and on the boundary of the designated spatial region). Given the rounding errors in computers, vortex shedding is naturally part of the solution to the Navier-Stokes equations. As both solvers are Navier-Stokes solvers, they were both expected to give a solution with vortex shedding. While the traditional CFD solver made it, the PINN solver did not in this paper.

SkillmanAnalyticsLLC replied:
thank you!

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+
+Figure :ref:`fig:cylinder-pinn-training-convergence` shows the convergence history.
+The bumps in the history correspond to our manual adjustment of the learning rates.
+After 64 hours of training, the total loss had not converged to an obvious steady value. 
```
My initial instinct would be to ask whether your model architecture/size is simply a poor choice. I (and many others) have certainly had the experience that something won't train at all because the model architecture is wrong. I wonder whether you have a response to this that you could include here (to justify why you don't think this is merely a function of a very wrong architecture). FC architectures can't be expected to model complex phenomena.

labarba replied:
The majority of papers on PINN use a fully connected architecture, because it is proved to be a universal approximator, as long as the network is big enough. Theoretically speaking, if you are not satisfied with the accuracy, you can increase the size of the network to improve it. Our paper did not have the goal of searching for a network architecture, or set of parameters that would _improve_ the results with PINN. Our goal is to report on the experience, from the point of view of CFD practitioners, when trying to use PINN for the first time.

piyueh replied:
To make the statement more complete: FC architectures are universal approximators for smooth functions. The solutions to the Navier-Stokes equation for the two flow cases in this paper are smooth. So as long as the network is big enough, the FC architectures are supposed to be able to give solutions that at least "look" correct. That's why using FC architectures is the first choice for us, and that's why the first thought we had when seeing PINN's incorrect solution was to increase the network's size.

SkillmanAnalyticsLLC replied:
I think it is helpful to know that the majority of papers on PINN use a fully connected architecture. It may be helpful to add this information to the report.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Allow us to start this discussion with a hypothetical situation.
+If one asks why we chose such a spatial and temporal resolution for a conventional CFD simulation, we have mathematical or physical reasons to back our decision.
+However, if the person asks why we chose 6 hidden layers and 256 neurons per layer, we will not be able to justify it.
+"It worked in another case!" is probably the best answer we can offer.
```
Did you cite another case where it worked? that would be good to know. But I think you make a good point that perhaps for PINNs in general there is not (yet) a well developed science, or even art, to developing architectures. So all this has shown is arguably that poor choice of architecture will get you nowhere and that (as you mention) there may not be many obvious systematic ways to improve. However, there are some obvious systematic ways to improve that you did not try (hyper parameter optimization, different architectures besides FC)

labarba replied:
@piyueh : why not add to the paper that the parameters come from the Modulus defaults, and they work for their examples

piyueh replied:
Done (https://github.com/barbagroup/scipy_proceedings/commit/f676e743746d29dfabd624f2409adef5140826ff)

piyueh replied:
I would like to clearify what we mean "systematic ways" in this paper. I think when you say "systematic way to improve", you mean something like grid-search, parameter sweep, or even something with more statistical meanings, like Taguchi methods. These are all practices of the [design of experiments (DoE)](https://en.wikipedia.org/wiki/Design_of_experiments). DoE relies on non-trivial attempts of runs to improve the parameters, which is not computationally affordable for us.

In traditional numerical methods, however, "systematic ways to improve solutions" frequently do not need many runs because most parameters are backed by mathematical derivations and analyses. For example, given the error bound and run time of a low-resolution CFD simulation, if now we want the error to go down two orders, we can estimate how fine the resolution should be and how long it's going to take by some simple algebraic calculations. We don't need to run several simulations just to figure out this information. This "estimating by math and theory" is what we mean when we say "systematic ways to improve solutions."

SkillmanAnalyticsLLC replied:
thank you.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Moreover, when it comes to something non-numeric, it is even more challenging to know what to use and why to use them.
+For instance, what activation function should we use and why?
+Should we use the same activation everywhere?
+Not to mention that we are not yet even considering a different network architecture here.
```
So you get to all my concerns here. I suppose I'm a little puzzled why you didn't try out some of these experiments.

labarba replied:
We didn't try those experiments because they would be too costly. One of the points we make in the paper is that (given the overly optimistic literature on the PINN method) using PINN for traditional CFD situations is tempting, but there's a long way to go before it's useful. We just don't want other students spending months trying this, and to learn from our experience.

piyueh replied:
If this paper is read as an experiment report, I think it may make more sense that we did not exhaust all possible configurations. The major aim of the paper is to report our experiments: what we have done, what we got, how we interpreted the results, and what we thought. It is not aiming to tackle any problems nor to make something work. Instead of publishing experimental data only after exhausting all possible configurations, we would like to share the results whenever we have new ones, especially failure results. I believe publishing failure results as soon as possible may save many other people's time from retrying the configurations that are not working. Hence I think not conducting thorough experiments is a limitation, but it is not possible to have a paper without limitations.

On the other hand, we did plan and are now trying some other configurations, but they will be in future publications.

SkillmanAnalyticsLLC replied:
Interesting. How would you plan to announce other failure (or non-failure) results. I agree that it's good to fight publication bias and to put out failure results as well as success results. Perhaps I missed this in the paper, but do you plan to perhaps consolidate results all in one place? or have some web reference that can at least list all experiments systematically undertaken in your various experiments?

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+Not to mention that we are not yet even considering a different network architecture here.
+
+Ultimately, are we even sure that increasing the network's complexity is the right path?
+Our assumption that the network is not complicated enough may just be wrong.
```
I think it could be. I'm especially thinking of something like NLP modeling where you won't get anywhere with FC. Or ditto for plenty of time series data.

piyueh replied:
I guess NLP problems do not have mathematically smooth and continuous solutions. That is to say, if there exists a powerful math equation that can answer an NLP problem, this equation is unlikely to be continuous and smooth. If so, then I think FC architectures cannot well approximate the solutions to NLP problems.

However, as mentioned previously, the solutions to this paper's flow problems are continuous and smooth, so theoretically FC should work. So I'm still not sure what happened.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+We were used to only having one set of initial conditions when the temporal derivative in differential equations is only first-order.
+However, in PINN, nothing limits us from using more than one initial condition.
+We can generate results at :math:`t=0,1,\dots,t_n` using a traditional CFD solver and add the residuals corresponding to these time snapshots to the total residual, so the PINN method may perform better in predicting :math:`t>t_n`.
+In other words, the PINN solver becomes traditional CFD solvers' replacement only for :math:`t>t_n`.
```
This whole paragraph is great, and I wish you would expand this both in your initial approach + the introduction. Maybe you could recast this a bit more as pitfalls of expecting CFD skills/methods to translate directly to PINN.

SkillmanAnalyticsLLC commented on papers/PiYueh_Chuang/main.rst:
```diff
@@ -0,0 +1,632 @@
+It may not work in many situations.
+Nevertheless, it does not mean we should stay away from PINN and discard this idea.
+Stepping away from a new thing makes zeros chance for it to evolve, and we will never know if PINN can be improved to a mature state that works well. 
+Of course, overly promoting its bright side with success stories does not help, either.
```
If there are specific cites where you think the authors are guilty of this, I think you should make those available.

labarba replied:
They are too many to list! And if we do name names, we could face some trolling, ha ha. Let's leave it as is.

SkillmanAnalyticsLLC replied:
ok!

stargaser commented:
@scoobies mark pending comment

deniederhut commented:
Thanks Aileen!

labarba committed:
fix typo

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899094375

labarba committed:
add pinn citation

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899095610

labarba committed:
fix word choice

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899100744

labarba committed:
add adverb

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899101191

labarba committed:
explain large computational graph

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899103806

labarba committed:
training points = spatio-temporal points

labarba committed:
add to: more GPUs not faster convergence

https://github.com/scipy-conference/scipy_proceedings/pull/705#discussion_r899120065

deniederhut commented:
Thanks for the revisions! Is this paper ready for a second read-through?

piyueh commented:
@deniederhut Not yet. When is the deadline? We will address the reviewer's comments and finish the revision today.

piyueh committed:
add note for weighted loss

piyueh committed:
add a paragraph to explain training data

piyueh committed:
add notice to the fact what we did are experiments, not analysis

piyueh committed:
add more references

deniederhut commented:
The deadline for the final 👍  from the reviewers is the 30th (https://github.com/scipy-conference/scipy_proceedings/tree/2022#timeline-for-2022), which is two days away. Typically this ends up bleeding into the week-end a bit, but you'll want to have everything ready asap so your reviewers have time to look things over, especially since most people will not want to work over a holiday week-end. 

(I will, of course, be here 😛)

SkillmanAnalyticsLLC commented:
I am happy to recommend this paper for the proceedings. Thank you @piyueh @labarba for the clarifications, explanations, and modifications!

stargaser commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/72.712178 is OK
- 10.2514/3.21122 is OK
- 10.2514/2.2134 is OK
- 10.1137/1.9780898719598 is OK
- 10.1017/S0022112080000419 is OK
- 10.1016/j.apm.2008.01.017 is OK
- 10.1063/1.2814259 is OK
- 10.21105/joss.00558 is OK
- 10.48550/ARXIV.1606.08415 is OK
- 10.1146/annurev.fl.28.010196.002401 is OK
- 10.1016/j.jcp.2018.10.045 is OK
- 10.5281/zenodo.6592457 is OK
- 10.1007/s10409-021-01148-1 is OK

MISSING DOIs

- 10.1109/sc41405.2020.00013 may be a valid DOI for title: MeshfreeFlowNet: A Physics-Constrained Deep Continuous Space-Time Super-Resolution Framework

INVALID DOIs

- 10.1016/S0893-60800300083-2 is INVALID
- 10.1016/0893-60809190009-T is INVALID
- 10.1016/0041-55537490061-5 is INVALID
```

stargaser commented:
@piyueh would you please take a look at the missing and invalid DOIs per the preceding comment?

piyueh committed:
replace an arXiv articla w/ its published version

piyueh commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/72.712178 is OK
- 10.2514/3.21122 is OK
- 10.2514/2.2134 is OK
- 10.1137/1.9780898719598 is OK
- 10.1017/S0022112080000419 is OK
- 10.1016/j.apm.2008.01.017 is OK
- 10.1063/1.2814259 is OK
- 10.21105/joss.00558 is OK
- 10.48550/ARXIV.1606.08415 is OK
- 10.1146/annurev.fl.28.010196.002401 is OK
- 10.1016/j.jcp.2018.10.045 is OK
- 10.1109/SC41405.2020.00013 is OK
- 10.5281/zenodo.6592457 is OK
- 10.1007/s10409-021-01148-1 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1016/S0893-60800300083-2 is INVALID
- 10.1016/0893-60809190009-T is INVALID
- 10.1016/0041-55537490061-5 is INVALID
```

piyueh commented:
Hi @stargaser, some DOIs have parentheses, but the `@scoobies` strips them:

* 10.1016/S0893-60800300083-2 is actually 10.1016/S0893-6080(03)00083-2
* 10.1016/0893-60809190009-T is actually 10.1016/0893-6080(91)90009-T
* 10.1016/0041-55537490061-5 is actually 10.1016/0041-5553(74)90061-5

How do I resolve it? Thanks.

stargaser commented:
Thank you for checking @piyueh. The scoobies check is not perfect so we should leave these DOIs as they are currently in the paper.

stargaser commented:
@scoobies mark ready

labarba commented:
@stargaser, @deniederhut — what is the next step? The reviewer recommended the paper for acceptance. I could do one last read-through for any remaining typos, and then you will process for minting a DOI?

piyueh committed:
revise abstract

deniederhut commented:
@labarba we will publish the papers marked "ready" in the conference-ready proceedings in a couple of days. The final publication and DOI generation will be at the end of the month when we also have all the submitted posters and slides ready to go.

piyueh committed:
revise introduction

labarba committed:
tighten text

labarba committed:
minor typos
